client.go

```diff
@@ -0,0 +1,60 @@
+package main
+
+import (
+	"bufio"
+	"net"
+
+	baps3 "github.com/UniversityRadioYork/baps3-go"
+)
+
+type Client struct {
+	conn       net.Conn
+	reader     *bufio.Reader
+	writer     *bufio.Writer
+	listenerCh chan []byte
+	Outgoing   chan []byte
+}
+
+func MakeClient(conn net.Conn, clientCh chan []byte) *Client {
+	client := &Client{
+		conn:       conn,
+		reader:     bufio.NewReader(conn),
+		writer:     bufio.NewWriter(conn),
+		listenerCh: clientCh,
+		Outgoing:   make(chan []byte),
+	}
+
+	go client.Read()
+	go client.write()
+
+	ohai := MakeWelcomeMsg()
+	data, _ := ohai.Pack()
+	client.Outgoing <- data
+
+	return client
+}
+
+func MakeWelcomeMsg() *baps3.Message {
+	return baps3.NewMessage(baps3.RsOhai).AddArg("listd").AddArg("0.0")
+}
+
+func (c *Client) Close() {
+	c.conn.Close()
+}
+
+func (c *Client) Read() {
+	for {
+		data, _ := c.reader.ReadBytes('\n')
+		c.listenerCh <- data // Each client doesn't care what it got, that's for the server to handle
+	}
+}
+
+func (c *Client) write() {
+	for data := range c.Outgoing {
+		_, err := c.writer.Write(data)
```

Comment:
I don't believe there is any benefit using a bufio.Writer here, as we have complete packed messages ready to go at once. Just call c.conn.Write(data) directly.